digdag-core/src/main/java/io/digdag/core/database/BasicDatabaseStoreManager.java

```diff
@@ -274,42 +283,105 @@ public int getIndex()
                             Throwables.propagateIfInstanceOf(ex, exClass);
                             Throwables.propagateIfPossible(ex);
                             throw new TransactionFailedException(
-                                    "Transaction failed do to exception being thrown " +
+                                    "Transaction failed due to exception being thrown " +
                                     "from within the callback. See cause " +
                                     "for the original exception.", ex);
                         }
-                        if (!ts.retryNext) {
-                            validateTransactionAndCommit(handle);
-                            return retval;
-                        }
-                        handle.rollback();
-                        ts.retryNext = false;
-                        ts.retryCount += 1;
+                        handle.commit();
```

Comment:
Hm...yes, maybe.
validateTransactionAndCommit takes effect when we need to rollback the transaction but inner action didn't throw an exception. Rollback is mandatory when a SQL statement fails. And usually, inner action throws an exception if a SQL statement fails. validateTransactionAndCommit can prevent unexpected behavior (commit is called but nothing is committed) when an inner action ignored SQL exceptions. But it has overhead (it issues a SELECT statement).

transactionWithLocalRetry always doesn't follow it (RetryableTransactionActionWithException) because it simply sets retry flag and doesn't throw exception when a SQL statement fails.